shared/libpng-util/libpng-util.cpp

```diff
@@ -0,0 +1,122 @@
+
+unsigned libpng_encode32_file(const unsigned char* image, const unsigned w, const unsigned h, const char* filename) {
+  FILE *fp = fopen(filename, "wb");
+
+  png_structp png = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+  if (!png) return -1;
+  png_infop info = png_create_info_struct(png);
+  if (!info) {
+    png_destroy_write_struct(&png, NULL);
+    return -2;
+  }
+
+  png_init_io(png, fp);
+
+  // Write header (8 bit colour depth)
+  png_set_IHDR(png, info, w, h,
+               8, PNG_COLOR_TYPE_RGBA, PNG_INTERLACE_NONE,
+               PNG_COMPRESSION_TYPE_BASE, PNG_FILTER_TYPE_BASE);
+
+  png_write_info(png, info);
+
+  for (unsigned i = 0; i < h; ++i) {
+    png_write_row(png, (png_bytep)&image[sizeof(png_byte) * 4 * w * i]);
+  }
+  png_write_end(png, NULL);
+
+  fclose(fp);
+  png_free_data(png, info, PNG_FREE_ALL, -1);
+  png_destroy_write_struct(&png, (png_infopp)NULL);
+
+  return 0;
+}
+
+unsigned libpng_decode32_file(unsigned char** out, unsigned* w, unsigned* h, const char* filename) {
+  (*w) = 0; (*h) = 0;
+  FILE *fp = fopen(filename, "rb");
+
+  png_structp png = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+  if (!png) return -1;
+  png_infop info = png_create_info_struct(png);
+  if (!info) {
+    png_destroy_write_struct(&png, NULL);
+    return -2;
+  }
+
+  png_init_io(png, fp);
+  png_read_info(png, info);
+
+  png_byte color_type, bit_depth;
+  (*w)       = png_get_image_width(png, info);
+  (*h)       = png_get_image_height(png, info);
+  color_type = png_get_color_type(png, info);
+  bit_depth  = png_get_bit_depth(png, info);
+
+  // Read any color_type into 8bit depth, RGBA format.
+  // See http://www.libpng.org/pub/png/libpng-manual.txt
+
+  if (bit_depth == 16)
+    png_set_strip_16(png);
+
+  if (color_type == PNG_COLOR_TYPE_PALETTE)
+    png_set_palette_to_rgb(png);
+
+  // PNG_COLOR_TYPE_GRAY_ALPHA is always 8 or 16bit depth.
+  if (color_type == PNG_COLOR_TYPE_GRAY && bit_depth < 8)
+    png_set_expand_gray_1_2_4_to_8(png);
+
+  if (png_get_valid(png, info, PNG_INFO_tRNS))
+    png_set_tRNS_to_alpha(png);
+
+  // These color_type don't have an alpha channel then fill it with 0xff.
+  if (color_type == PNG_COLOR_TYPE_RGB ||
+      color_type == PNG_COLOR_TYPE_GRAY ||
+      color_type == PNG_COLOR_TYPE_PALETTE)
+    png_set_filler(png, 0xFF, PNG_FILLER_AFTER);
+
+  if (color_type == PNG_COLOR_TYPE_GRAY ||
+      color_type == PNG_COLOR_TYPE_GRAY_ALPHA)
+    png_set_gray_to_rgb(png);
+
+  png_read_update_info(png, info);
+
+  png_bytep image;
+  size_t pitch = sizeof(png_byte) * 4 * (*w); // number of bytes in a row
+  image = (png_bytep)malloc(pitch * (*h));
+
+  for (size_t y = 0; y < (*h); y++) {
+    png_read_row(png, (png_bytep)&image[pitch * y], NULL);
+  }
+
+  png_destroy_read_struct(&png, &info, NULL);
+  fclose(fp);
+
+  (*out) = image;
+
+  return 0;
+}
+
+std::string libpng_error_text(unsigned error) {
+  return "";
```

Comment:
wut

Reply:
Yeah, previously this was `lodepng_error_text` and we were using it to print a more specific error. I couldn't find anything in libpng to return such an error message here, so I left it. Would you like it to be removed?

Reply:
"wut" means "remove it", Robert.

Reply:
Sure, but I was wondering if he had a better inkling than I of what to replace it with before I preemptively removed it.

Reply:
So as an example, `"Error code " + std::tostring(error)` is infinitely better than what we have. You could also actually look up the PNG error string.

Reply:
Ok: 46814b8d58c1b2af9a123c856fe215d28af38418

So, the thing is, libpng is different than LodePNG and you don't need to ask it for the error message from a code. Its default error handling just prints to the standard error stream, and the way you do custom error handling is by providing your own error handling function. So for now, I just removed the stuff.

http://refspecs.linuxbase.org/LSB_3.1.0/LSB-Desktop-generic/LSB-Desktop-generic/libpng12.png.create.read.struct.1.html